api/cas-server-core-api-mfa/src/main/java/org/apereo/cas/authentication/MultifactorAuthenticationProviderFactory.java

```diff
@@ -0,0 +1,36 @@
+package org.apereo.cas.authentication;
+
+import org.apereo.cas.configuration.model.support.mfa.BaseMultifactorProviderProperties;
+
+/**
+ * Interface for provider a factory that can create a MultifactorAuthenticationProvider.
+ *
+ * @author Travis Schmidt
+ * @since 6.0
+ */
+public interface MultifactorAuthenticationProviderFactory<T extends MultifactorAuthenticationProvider,
+                                                          P extends BaseMultifactorProviderProperties> {
+
+    /**
+     * Default suffix to add to the bean name.
+     */
+    String PROVIDER_SUFFIX = "-provider";
+
+    /**
+     * Create an instance of MultifactorAuthenticationProvider based on passed properties.
+     *
+     * @param properties - the properties
+     * @return - the provider
+     */
+    T create(P properties);
+
+    /**
+     * Method returns the generated bean name for the provider.
+     *
+     * @param id - the id
+     * @return - the beanName
+     */
+    default String beanName(final String id) {
```

Comment:
I went down that hole too, and that was what the previous mess with creating BeanDefinitions used to create the beans.  The issue with using AbstractFactoryBean is:

1. Comes with a lot of Spring baggage.
2. createInstance() or getObject() depending if you use the abstract class or interface does not take parameters and relies on properties set in the bean at creation. 
3. Because of 2 you have to configure bean definitions on the fly, and if you want it refresh, it gets ugly.

The intent of the bean is to be an overridable interface that end deployers can customize if needed for their providers.  Would some other term besides "Factory" resolve potential confusion?

Reply:
Understood. I think if you then switch to `FactoryBean` then it's quite clear what this thing is supposed to do. And do please document the caveats in javadocs so the usage and decisions made are clear.

Reply:
You mean just add "Bean" to the end of the class name and not revert back to previously agreed code that induced vomiting on sight?

Reply:
Yes :)